modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py

```diff
@@ -93,76 +123,62 @@ def apply(self, func, *args, **kwargs):
         -----
         It does not matter if `func` is callable or an ``ray.ObjectRef``. Ray will
         handle it correctly either way. The keyword arguments are sent as a dictionary.
+
+        If ``LazyExecution`` is enabled, the function is not applied immediately,
+        but is added to the execution tree.
         """
-        log = get_logger()
-        self._is_debug(log) and log.debug(f"ENTER::Partition.apply::{self._identity}")
-        data = self._data
-        call_queue = self.call_queue + [[func, args, kwargs]]
-        if len(call_queue) > 1:
-            self._is_debug(log) and log.debug(
-                f"SUBMIT::_apply_list_of_funcs::{self._identity}"
-            )
-            result, length, width, ip = _apply_list_of_funcs.remote(
-                data, *deconstruct_call_queue(call_queue)
+        if LazyExecution.get():
+            de = DeferredExecution(self._data_ref, func, args, kwargs)
+            return self.__constructor__(de)
+
+        data = self._data_ref
+        if not isinstance(data, DeferredExecution):
+            flat_args = not has_list_or_de(args)
+            flat_kwargs = not has_list_or_de(kwargs.values())
+            if flat_args and flat_kwargs:
+                result = remote_exec_func.remote(func, data, *args, **kwargs)
+                return self.__constructor__(*result)
+            de = DeferredExecution(
+                data, func, args, kwargs, flat_args, flat_kwargs, num_returns=1
             )
         else:
-            # We handle `len(call_queue) == 1` in a different way because
-            # this dramatically improves performance.
-            func, f_args, f_kwargs = call_queue[0]
-            result, length, width, ip = _apply_func.remote(
-                data, func, *f_args, **f_kwargs
-            )
-            self._is_debug(log) and log.debug(f"SUBMIT::_apply_func::{self._identity}")
-        self._is_debug(log) and log.debug(f"EXIT::Partition.apply::{self._identity}")
-        return self.__constructor__(result, length, width, ip)
+            de = DeferredExecution(data, func, args, kwargs)
+        de.ref_count(1)
```

Comment:
Why?

Reply:
If we don't have refs, we will not get the result.

Reply:
It’s a little strange that in order to perform `exec` function, we need to separately increase the ref count. Maybe we should move this to `DeferredExecution.__init__`?

Reply:
Done.